dataflow/src/main/java/org/checkerframework/dataflow/analysis/ForwardAnalysisImpl.java

```diff
@@ -0,0 +1,585 @@
+import org.checkerframework.dataflow.cfg.UnderlyingAST;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.CFGLambda;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.CFGMethod;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.Kind;
+import org.checkerframework.dataflow.cfg.block.Block;
+import org.checkerframework.dataflow.cfg.block.ConditionalBlock;
+import org.checkerframework.dataflow.cfg.block.ExceptionBlock;
+import org.checkerframework.dataflow.cfg.block.RegularBlock;
+import org.checkerframework.dataflow.cfg.block.SpecialBlock;
+import org.checkerframework.dataflow.cfg.node.LocalVariableNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.ReturnNode;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.Pair;
+
+/**
+ * An implementation of a forward analysis to solve a org.checkerframework.dataflow problem given a
+ * control flow graph and a forward transfer function.
+ *
+ * @param <V> the abstract value type to be tracked by the analysis
+ * @param <S> the store type used in the analysis
+ * @param <T> the transfer function type that is used to approximate runtime behavior
+ */
+public class ForwardAnalysisImpl<
+                V extends AbstractValue<V>,
+                S extends Store<S>,
+                T extends ForwardTransferFunction<V, S>>
+        extends AbstractAnalysis<V, S, T> implements ForwardAnalysis<V, S, T> {
+
+    /**
+     * Number of times each block has been analyzed since the last time widening was applied. Null
+     * if maxCountBeforeWidening is -1, which implies widening isn't used for this analysis.
+     */
+    protected final @Nullable IdentityHashMap<Block, Integer> blockCount;
+
+    /**
+     * Number of times a block can be analyzed before widening. -1 implies that widening shouldn't
+     * be used.
+     */
+    protected final int maxCountBeforeWidening;
+
+    /** Then stores before every basic block (assumed to be 'no information' if not present). */
+    protected final IdentityHashMap<Block, S> thenStores;
+
+    /** Else stores before every basic block (assumed to be 'no information' if not present). */
+    protected final IdentityHashMap<Block, S> elseStores;
+
+    /** The stores after every return statement. */
+    protected final IdentityHashMap<ReturnNode, TransferResult<V, S>> storesAtReturnStatements;
+
+    // `@code`, not `@link`, because dataflow module doesn't depend on framework module.
+    /**
+     * Construct an object that can perform a org.checkerframework.dataflow forward analysis over a
+     * control flow graph. The transfer function is set by the subclass, e.g., {@code
+     * org.checkerframework.framework.flow.CFAbstractAnalysis}, later.
+     *
+     * @param maxCountBeforeWidening number of times a block can be analyzed before widening
+     */
+    public ForwardAnalysisImpl(int maxCountBeforeWidening) {
```

Comment:
Why does the forward analysis contain widening support, but the backwards analysis does not? I know that there is no need for a widening operator in the liveness analysis, but to be generally useful I would expect that there will be times when a widening operator is needed - otherwise a dataflow problem might not terminate. The backwards analysis is not immune to this (as far as I can tell); it must re-add blocks to the worklist just as the forward analysis does. Am I missing something, or is widening support in the backwards analysis future work?